File: Leetcode/tree/894_AllPossibleFullBinaryTrees/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<TreeNode> allPossibleFBT(int N) {
        List<TreeNode> res = new ArrayList<>();
        if(N==1){
            res.add(new TreeNode(0));
            return res;
        }
        N-=1;
        for(int i = 1; i<N; i+=2){
            List<TreeNode> left = allPossibleFBT(i);
            List<TreeNode> right = allPossibleFBT(N-i);
            for(TreeNode l: left){
                for(TreeNode r: right){
                    TreeNode cur = new TreeNode(0);
                    cur.left = l;
                    cur.right = r;
                    res.add(cur);
                }
            }
        }
        return res;
    }
}

File: Leetcode/tree/894_AllPossibleFullBinaryTrees/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def allPossibleFBT(self, N):
        """
        :type N: int
        :rtype: List[TreeNode]
        """
        res = []
        if N==1:
            res.append(TreeNode(0))
            return res
        N-=1
        for i in xrange(1, N, 2):
            left = self.allPossibleFBT(i)
            right = self.allPossibleFBT(N-i)
            for l in left:
                for r in right:
                    cur = TreeNode(0)
                    cur.left = l
                    cur.right = r
                    res.append(cur)
        return res
